v2/block_reader.go

```diff
@@ -122,5 +129,88 @@ func (br *BlockReader) Next() (blocks.Block, error) {
 		return nil, fmt.Errorf("mismatch in content integrity, expected: %s, got: %s", c, hashed)
 	}
 
+	ss := uint64(c.ByteLen()) + uint64(len(data))
+	br.offset += uint64(varint.UvarintSize(ss)) + ss
 	return blocks.NewBlockWithCid(data, c)
 }
+
+type BlockMetadata struct {
+	cid.Cid
+	Offset uint64
+	Size   uint64
+}
+
+// SkipNext jumps over the next block, returning metadata about what it is (the CID, offset, and size).
+// Like Next it will return an io.EOF once it has reached the end.
+//
+// If the underlying reader used by the BlockReader is actually a ReadSeeker, this method will attempt to
+// seek over the underlying data rather than reading it into memory.
+func (br *BlockReader) SkipNext() (*BlockMetadata, error) {
+	sctSize, err := util.LdReadSize(br.r, br.opts.ZeroLengthSectionAsEOF, br.opts.MaxAllowedSectionSize)
+	if err != nil {
+		return nil, err
+	}
+
+	if sctSize == 0 {
+		_, _, err := cid.CidFromBytes([]byte{})
+		return nil, err
+	}
+
+	cidSize, c, err := cid.CidFromReader(io.LimitReader(br.r, int64(sctSize)))
+	if err != nil {
+		return nil, err
+	}
+
+	blkSize := sctSize - uint64(cidSize)
+
+	if brs, ok := br.r.(io.ReadSeeker); ok {
+		if br.readerSize == -1 {
+			cur, err := brs.Seek(0, io.SeekCurrent)
+			if err != nil {
+				return nil, err
+			}
+			end, err := brs.Seek(0, io.SeekEnd)
+			if err != nil {
+				return nil, err
+			}
+			br.readerSize = end
+			if _, err = brs.Seek(cur, io.SeekStart); err != nil {
+				return nil, err
+			}
+		}
+		// seek.
+		finalOffset, err := brs.Seek(int64(blkSize), io.SeekCurrent)
+		if err != nil {
+			return nil, err
+		}
+		if finalOffset != int64(br.offset)+int64(sctSize)+int64(varint.UvarintSize(sctSize)) {
+			return nil, fmt.Errorf("unexpected length")
+		}
+		if finalOffset > br.readerSize {
+			return nil, io.ErrUnexpectedEOF
+		}
+		br.offset = uint64(finalOffset)
+		return &BlockMetadata{
+			c,
+			uint64(finalOffset) - sctSize - uint64(varint.UvarintSize(sctSize)),
+			blkSize,
+		}, nil
+	}
+
+	// read to end.
+	_, err = io.CopyN(io.Discard, br.r, int64(blkSize))
```

Comment:
Any reason to not check the `written` value against expected, like we're doing for the seek version above?

Reply:
copy-to-discard should be able to not allocate a buffer when moving the reader forward

Reply:
but it still returns a number of bytes "copied", right? so checking that the return value matches `blkSize` seems like an appropriate thing to do?